### PR TITLE
enable llava static generation.

### PR DESCRIPTION
**Pull request by lkk12014402**
# What does this PR do?

support llava image to text generation

### CODE REVIEW

**Comment by lkk12014402** [comment]
based on the image-to-text generation pr [#738](https://github.com/huggingface/optimum-habana/pull/738)

```
python3 run_pipeline.py \
        --model_name_or_path "llava-hf/llava-1.5-7b-hf" \
        --image_path "https://llava-vl.github.io/static/images/view.jpg" \
        --prompt "<image>\nUSER: What's the content of the image?\nASSISTANT:" \
        --max_new_tokens 20 \
        --use_hpu_graphs \
        --bf16
```
```
result = [[{'generated_text': "[\\nUSER: What's the content of the image?\\nASSISTANT: The image features a pier extending out into a large body of water, likely a lake.\n\n](file://nUSER:%20What's%20the%20content%20of%20the%20image%3f/nASSISTANT:%20The%20image%20features%20a%20pier%20extending%20out%20into%20a%20large%20body%20of%20water,%20likely%20a%20lake./n/n)"}]], time = 264.1947269439697ms
```

**Comment by lkk12014402** [comment]
- For batch_size = 4
Input/outputs 1:
USER: What's the content of the image?
ASSISTANT: The image features a pier extending out into a large body of water, likely a lake. The pier
Input/outputs 2:
USER:  What's the content of the image?
ASSISTANT: The image features a pier extending out over a large body of water, likely a lake. The pier
Input/outputs 3:
USER: describe the image?
ASSISTANT: The image features a pier extending out into a large body of water, likely a lake. The pier
Input/outputs 4:
USER: Is there a brige in the image?
ASSISTANT: Yes, there is a bridge in the image.
USER: Is the bridge over water?

Input/outputs:
Throughput (including tokenization) = 191.2782859461739 tokens/second

Number of HPU graphs                = 26
Memory allocated                    = 16.26 GB
Max memory allocated                = 16.27 GB
Total memory available              = 94.62 GB

**Comment by JoeyTPChou** [comment]
Just want to let you know this works like a charm!

**Comment by JoeyTPChou** [comment]
Hi @lkk12014402 I tried with no image and looks like it doesn't support input with no image. Want to check if this is correct and see if we can support no image input (text only) ?

**Review by ssarkar2 — COMMENTED** [review]
@lkk12014402 , could you please provide a brief description of the changes needed in `optimum/habana/transformers/models/llava/modeling_llava.py` wrt the base model in transformers

**Comment by lkk12014402** [comment]
> @lkk12014402 , could you please provide a brief description of the changes needed in `optimum/habana/transformers/models/llava/modeling_llava.py` wrt the base model in transformers
> 
> I see a couple of single input `where`, which are usually dynamic on HPU. If these are on CPU, then its fine, but if these are on HPU, they might need rewriting.
> 
> https://github.com/huggingface/optimum-habana/blob/d44c54066aef31462206a4d3d4a4b4e93a72dd01/optimum/habana/transformers/models/llava/modeling_llava.py#L79
> 
> 
> https://github.com/huggingface/optimum-habana/blob/d44c54066aef31462206a4d3d4a4b4e93a72dd01/optimum/habana/transformers/models/llava/modeling_llava.py#L29

hi, @ssarkar2 

I will give a description and check the operation torch.where() as soon as possible

**Comment by lkk12014402** [comment]
> @lkk12014402 , could you please provide a brief description of the changes needed in `optimum/habana/transformers/models/llava/modeling_llava.py` wrt the base model in transformers
> 
> I see a couple of single input `where`, which are usually dynamic on HPU. If these are on CPU, then its fine, but if these are on HPU, they might need rewriting.
> 
> https://github.com/huggingface/optimum-habana/blob/d44c54066aef31462206a4d3d4a4b4e93a72dd01/optimum/habana/transformers/models/llava/modeling_llava.py#L79
> 
> 
> https://github.com/huggingface/optimum-habana/blob/d44c54066aef31462206a4d3d4a4b4e93a72dd01/optimum/habana/transformers/models/llava/modeling_llava.py#L29

hi, @ssarkar2 ,

#### Description
Let's assume the input text is ["hey" "<image>", "how", "are"], and one image

##### generation with huggingface transformers directly
the huggingface transformers will get the text embedding [1, 4, 4096] with [llava-1.5-7b-hf](https://huggingface.co/llava-hf/llava-1.5-7b-hf), and get image embedding [1,576, 4096]. Then the 2 embeddings will be merged to final input embedding [1, 579, 4096] using [here](https://github.com/huggingface/transformers/blob/main/src/transformers/models/llava/modeling_llava.py#L431). 

The merge function also has many dynamic op, like torch.where and the input shape is dynamic during the generation. 

So when we use gaudi2 to do generation, there are 2 problems:
1. the generation is very slow, and we have test one example, like [this](https://github.com/huggingface/transformers/blob/main/src/transformers/models/llava/modeling_llava.py#L382)
```
python3 run_pipeline.py \
    --model_name_or_path llava-hf/llava-1.5-7b-hf \
    --image_path https://llava-vl.github.io/static/images/view.jpg \
    --prompt "<image>\nUSER: What's the content of the image?\nASSISTANT:" \
    --max_new_tokens 20 \
--bf16
 
Output is 
03/04/2024 05:22:07 - INFO - __main__ - result = [[{'generated_text': \\nUSER: What's the content of the image?\\nASSISTANT: The image features a pier extending out into a large body of water, likely a lake.\n\n}]], time = 1148.6382484436035ms
```
**note:** to reproduce, you can use [this pr](https://github.com/huggingface/optimum-habana/pull/738) image-to-text example

2. we can not apply --use_hpu_graph, because there are some errors.

##### my optimization

In order to maintain the transformers usage (same input, same generation script) and enable static shape by padding and inserting token_idx for generation, I add a new function `_pad_inputs` to pad input. The function will extend the special token `<image>` to the num patches of image feature, the padded input text can be regarded as ["hey" "<image>", "<image>", ......, "<image>""how", "are"] which sequence length is 579. So the text embedding shape is [1, 579, 4096]. When we merge 2 embeddings (text embedding and image embedding), we don't need complex compute like [this](https://github.com/huggingface/transformers/blob/main/src/transformers/models/llava/modeling_llava.py#L279) and I simplified the function as you see in the `modeling_llava.py` file.

And for keeping same input shape during generation, I also use token_idx. So I create 2 auxiliary variables,  `tokens_pos` and `image_offset`. The `tokens_pos` records the original input text position to select logits, which can keep same shape between input text/ids and output logits. The `image_offset` records the offset because of the special tokens <image>, which should be added to the token_idx during the model forward.


#### the explanation of maintaining `torch.where`

we need use this function to compute the special token index, because we don't preprocess the input.

Ideally, we should preprocess the input, like the padding and extending the special tokens `<image>`, which will need more changes compared to transformers, especially we need to create a new generation script instead of using the script in [this pr](https://github.com/huggingface/optimum-habana/pull/738).

**Comment by lkk12014402** [comment]
@ssarkar2 please help review~

**Review by libinta — COMMENTED** [review]
@lkk12014402 can you add a ci test case and rebase?

**Comment by lkk12014402** [comment]
> @lkk12014402 can you add a ci test case and rebase?

@libinta I will update the pr with your comments soon.

**Comment by libinta** [comment]
@lkk12014402  can you add a file like test_image2text_generation_example.py to include image2text generation
and change https://github.com/huggingface/optimum-habana/blob/081130d53a1237e5b8b5d7ac1c64195300579303/Makefile#L76 to include it?

**Comment by lkk12014402** [comment]
> @lkk12014402 can you add a file like test_image2text_generation_example.py to include image2text generation and change
> 
> https://github.com/huggingface/optimum-habana/blob/081130d53a1237e5b8b5d7ac1c64195300579303/Makefile#L76
> 
> to include it?

hi @libinta  please help review/check the image to text ut. Thanks~

File: tests/test_image_to_text_example.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+        with open(Path(tmp_dir) / "results.json") as fp:
+            results = json.load(fp)
+
+        # Ensure performance requirements (throughput) are met
+        assert results["throughput"] >= (2 - TIME_PERF_FACTOR) * baseline
+
+
+@pytest.mark.parametrize("model_name, batch_size, reuse_cache, baseline", MODELS_TO_TEST["bf16"])
+def test_text_generation_bf16(model_name: str, baseline: float, batch_size: int, reuse_cache: bool, token: str):
```
Comment (libinta) [inline tests/test_image_to_text_example.py:88]: better to have image_to_test rather than text_generation

File: tests/test_image_to_text_example.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    command = ["python3"]
+    path_to_example_dir = Path(__file__).resolve().parent.parent / "examples"
+    env_variables = os.environ.copy()
+
+    command += [
+        f"{path_to_example_dir / 'image-to-text' / 'run_pipeline.py'}",
+        f"--model_name_or_path {model_name}",
+        f"--batch_size {batch_size}",
+        "--use_kv_cache",
+        "--max_new_tokens 20",
```
Comment (libinta) [inline tests/test_image_to_text_example.py:41]: have you ran the test with
 GAUDI2_CI=1 RUN_SLOW=true  python -m pytest tests/test_image_to_text_example.py -v -s
if so, you will see run_pipeline.py: error: unrecognized arguments: --use_kv_cache --output_dir /tmp/tmpsp9f6li_ --token None
you should include whatever arguments as  python3 run_pipeline.py \
        --model_name_or_path "llava-hf/llava-1.5-7b-hf" \
        --image_path "https://llava-vl.github.io/static/images/view.jpg" \
        --prompt "<image>\nUSER: What's the content of the image?\nASSISTANT:" \
        --max_new_tokens 20 \
        --use_hpu_graphs \
        --bf16

File: tests/test_image_to_text_example.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    with TemporaryDirectory() as tmp_dir:
+        command.append(f"--output_dir {tmp_dir}")
+        print(f"\n\nCommand to test: {' '.join(command)}\n")
+
+        command.append(f"--token {token.value}")
+
+        pattern = re.compile(r"([\"\'].+?[\"\'])|\s")
+        command = [x for y in command for x in re.split(pattern, y) if x]
+
+        if fp8:
```
Comment (libinta) [inline tests/test_image_to_text_example.py:59]: remove fp8 for now

File: optimum/habana/transformers/models/llava/modeling_llava.py [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+                # past_length += token_idx
+                input_ids = torch.index_select(input_ids, 1, token_idx + image_offset - 1)
+
+        position_ids = kwargs.get("position_ids", None)
+        if attention_mask is not None and position_ids is None:
+            # create position_ids on the fly for batch generation
+            position_ids = attention_mask.long().cumsum(-1) - 1
+            position_ids.masked_fill_(attention_mask == 0, 1)
+            if past_key_values:
+                position_ids = torch.sum(attention_mask, dim=1).unsqueeze(-1) - 1
```
Comment (regisss) [inline optimum/habana/transformers/models/llava/modeling_llava.py:282]: It seems to be different here: https://github.com/huggingface/transformers/blob/v4.38.2/src/transformers/models/llava/modeling_llava.py#L544

**Review by regisss — COMMENTED** [review]
It seems there are some merge conflicts to solve, can you update your main branch and merge it into this one?
Also, please run
```
pip install -U ruff
make style
```
to have the code style check pass.

File: tests/test_image_to_text_example.py (hunk 1 above)
Comment (lkk12014402) [reply tests/test_image_to_text_example.py]: fixed

File: tests/test_image_to_text_example.py (hunk 2 above)
Comment (lkk12014402) [reply tests/test_image_to_text_example.py:34]: updated

File: tests/test_image_to_text_example.py (hunk 3 above)
Comment (lkk12014402) [reply tests/test_image_to_text_example.py]: removed

**Comment by lkk12014402** [comment]
> It seems there are some merge conflicts to solve, can you update your main branch and merge it into this one? Also, please run
> 
> ```
> pip install -U ruff
> make style
> ```
> 
> to have the code style check pass.

update code style with the command

File: optimum/habana/transformers/models/llava/modeling_llava.py [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+
+import torch
+from typing import List, Optional, Tuple, Union
+from transformers.cache_utils import Cache
+from transformers.models.llava.modeling_llava import LlavaForConditionalGeneration, LlavaCausalLMOutputWithPast
+from transformers.utils import logging
+
+logger = logging.get_logger(__name__)
+
+def _pad_inputs(input_ids, attention_mask, image_token_index, num_patches,
```
[the remark this answers is not in the document]
Comment (lkk12014402) [reply optimum/habana/transformers/models/llava/modeling_llava.py]: there is no padding function for image + text, so we need to preprocess them before `prepare_inputs_for_generation`

[the remark this answers is not in the document]
Comment (lkk12014402) [reply optimum/habana/transformers/models/llava/modeling_llava.py]: I implemented the padding func by myself, no reference

File: optimum/habana/transformers/models/llava/modeling_llava.py [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+        tokens_pos_padded.append([x + pad_len for x in cur_token_pos])
+
+    input_ids = torch.tensor(new_input_ids_padded).to(input_ids.device)
+    attention_mask = torch.tensor(new_attention_mask_padded).to(input_ids.device)
+    tokens_pos = torch.tensor(tokens_pos_padded).to(input_ids.device)
+
+    return input_ids, attention_mask, image_offset, tokens_pos
+
+
+def _merge_input_ids_with_image_features(image_features, inputs_embeds, input_ids, image_token_index):
```
[the remark this answers is not in the document]
Comment (lkk12014402) [reply optimum/habana/transformers/models/llava/modeling_llava.py:94]: the `_merge_input_ids_with_image_features` here [transformers/models/llava/modeling_llava.py#L279](https://github.com/huggingface/transformers/blob/v4.38.2/src/transformers/models/llava/modeling_llava.py#L279) is a little complicated and some code are not friendly for gaudi2. So with the rewritten `_pad_inputs`, the implementation can be simplified as I wrote like this.

File: optimum/habana/transformers/models/llava/modeling_llava.py (hunk 4 above)
Comment (lkk12014402) [reply optimum/habana/transformers/models/llava/modeling_llava.py]: updated when `token_idx==None`

File: examples/image-to-text/run_pipeline.py [hunk 7]
```diff
@@ -67,12 +69,23 @@ def main():
...
+        type=str,
+        help="Output directory to store results in.",
+    )
     parser.add_argument("--batch_size", type=int, default=1, help="Input batch size.")
     parser.add_argument("--warmup", type=int, default=3, help="Number of warmup iterations for benchmarking.")
     parser.add_argument("--n_iterations", type=int, default=5, help="Number of inference iterations for benchmarking.")
     args = parser.parse_args()
 
     adapt_transformers_to_gaudi()
+    if args.image_path is None and "llava" in args.model_name_or_path:
```
Comment (regisss) [inline examples/image-to-text/run_pipeline.py:84]: If we want to do this for all Llava models (i.e. also Llava models that don't have "llava" in their name), we should do:
```
from transformers import AutoConfig
model_type = AutoConfig.from_pretrained(args.model_name_or_path).model_type
if args.image_path is None and model_type == "llava":
```

File: examples/image-to-text/run_pipeline.py [hunk 8]
```diff
@@ -67,12 +69,23 @@ def main():
...
+        args.image_path = ["https://llava-vl.github.io/static/images/view.jpg"]
+    if args.prompt is None and "llava" in args.model_name_or_path:
```
Comment (regisss) [inline examples/image-to-text/run_pipeline.py:86]: Same

File: examples/image-to-text/run_pipeline.py (hunk 7 above)
Comment (lkk12014402) [reply examples/image-to-text/run_pipeline.py]: fixed

File: examples/image-to-text/run_pipeline.py (hunk 8 above)
Comment (lkk12014402) [reply examples/image-to-text/run_pipeline.py]: updated

**Comment by lkk12014402** [comment]
hi, @regisss updated code with your comments.
And the ut results:

```
04/24/2024 15:19:05 - INFO - __main__ - result = [[{'generated_text': "\nUSER: What's the content of the image?\nASSISTANT: The image features a pier extending out into a large body of water, likely a lake. The pier"}]], time = 245.72740799630992ms, Throughput (including tokenization) = 81.39100218035239 tokens/second
PASSED
```

**Review by regisss — APPROVED** [review]
